wallets/dcr/sync.go

```diff
@@ -0,0 +1,484 @@
+package dcr
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"sort"
+	"strings"
+	"sync"
+
+	"decred.org/dcrwallet/v2/errors"
+	"decred.org/dcrwallet/v2/p2p"
+	w "decred.org/dcrwallet/v2/wallet"
+	"github.com/decred/dcrd/addrmgr/v2"
+	"github.com/planetdecred/dcrlibwallet/spv"
+)
+
+// reading/writing of properties of this struct are protected by mutex.x
+type SyncData struct {
+	mu sync.RWMutex
+
+	SyncProgressListeners map[string]SyncProgressListener
+	showLogs              bool
+
+	synced       bool
+	syncing      bool
+	cancelSync   context.CancelFunc
+	cancelRescan context.CancelFunc
+	syncCanceled chan struct{}
+
+	// Flag to notify syncCanceled callback if the sync was canceled so as to be restarted.
+	restartSyncRequested bool
+
+	rescanning     bool
+	connectedPeers int32
+
+	*activeSyncData
+}
+
+// reading/writing of properties of this struct are protected by syncData.mu.
+type activeSyncData struct {
+	syncer *spv.Syncer
+
+	syncStage int32
+
+	cfiltersFetchProgress    CFiltersFetchProgressReport
+	headersFetchProgress     HeadersFetchProgressReport
+	addressDiscoveryProgress AddressDiscoveryProgressReport
+	headersRescanProgress    HeadersRescanProgressReport
+
+	addressDiscoveryCompletedOrCanceled chan bool
+
+	rescanStartTime int64
+
+	totalInactiveSeconds int64
+}
+
+const (
+	InvalidSyncStage          = -1
+	CFiltersFetchSyncStage    = 0
+	HeadersFetchSyncStage     = 1
+	AddressDiscoverySyncStage = 2
+	HeadersRescanSyncStage    = 3
+)
+
+func (wallet *Wallet) initActiveSyncData() {
+
+	cfiltersFetchProgress := CFiltersFetchProgressReport{
+		GeneralSyncProgress:         &GeneralSyncProgress{},
+		beginFetchCFiltersTimeStamp: 0,
+		startCFiltersHeight:         -1,
+		cfiltersFetchTimeSpent:      0,
+		totalFetchedCFiltersCount:   0,
+	}
+
+	headersFetchProgress := HeadersFetchProgressReport{
+		GeneralSyncProgress:      &GeneralSyncProgress{},
+		beginFetchTimeStamp:      -1,
+		headersFetchTimeSpent:    -1,
+		totalFetchedHeadersCount: 0,
+	}
+
+	addressDiscoveryProgress := AddressDiscoveryProgressReport{
+		GeneralSyncProgress:       &GeneralSyncProgress{},
+		addressDiscoveryStartTime: -1,
+		totalDiscoveryTimeSpent:   -1,
+	}
```

Comment:
Where addition is conducted later on this `-1`  default values should be compensated later on to avoid misrepresenting the actual figures.